### ISSUE TITLE
Does not the rpmbuild command execute (de)compression of files in parallel?

### ISSUE THREAD

**Issue by leemgs** [rpm-software-management/rpm#113]
It seems that a 'rpmbuild' command only support single-core cpu environment. Can we decompress and compress the binary files  in parallel while executing a rpm packaging operation with 'rpmbuild' command? 

 * The log messages while executing 'rpmbuild' command
```
 
[   46s] + exec rpmbuild --define '_srcdefattr (-,root,root)' --nosignature --target=armv7l-tizen-linux --define '_build_create_debug 1' -ba /home/abuild/rpmbuild/SOURCES/chromium-efl.spec
[   46s] Building target platforms: armv7l-tizen-linux
[   46s] Building for target armv7l-tizen-linux
[   46s] Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.nfcsUa
[   46s] + umask 022
[   46s] + cd /home/abuild/rpmbuild/BUILD
[   46s] + cd /home/abuild/rpmbuild/BUILD
[   46s] + rm -rf chromium-efl-47.2526.69.49
[   46s] + /bin/bzip2 -dc /home/abuild/rpmbuild/SOURCES/chromium-efl-47.2526.69.49.tar.bz2
[   46s] + /bin/tar -xf -
[  161s] + STATUS=0
[  161s] + '[' 0 -ne 0 ']'
[  161s] + cd chromium-efl-47.2526.69.49
[  161s] + /bin/chmod -Rf a+rX,u+w,g-w,o-w .
[  164s] + exit 0
[  164s] Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.JOaBJs
```
As we can see, the rpmbuild uses out-of-date bzip command without  up-to-date  pbzip command. 
Anyone who had similar experience like me in multi-core CPU(s)? 

In the past, we sometimes use a "--threads" option of xz package. For example, "$ tar -cf - source | xz --threads=0 > foo-destination.tar.xz [enter]".  Currently, It seems that we can just enable multi-threaded operation with xz(=lzma) compression type for rpm packaging in multi-core systems according to the rpm mailing-list (lists.rpm.org) as following:

* rpm-maint: MT compression in rpmbuild ?: http://lists.rpm.org/pipermail/rpm-maint/2015-March/003869.html

* rpm version 4.8.1's changelog: http://upstream.rosalinux.ru/changelogs/librpm4/4.8.1/changelog.html
commit edc5c0ef84a93d19acecaaf9a07f4ea3e9d8fe39 
commit 341266ae26e6685d1d870b28533469c653355bb7 

* github.com: Add support for multithreaded xz compression
https://github.com/rpm-software-management/rpm/commit/7740d1098642cd42f725fb9a2a3819ceaf51a80b#commitcomment-14626193

**Comment by leemgs** [comment]
As a workaround, How about we try to do the below way? At this time, Let's assume that the default compression of rpm package is "*.tar.bz2".
```
u$> cat ./sample.spec
  1 Summary: The "Hello World" program from GNU                                                                        
  2 Name: hello
  3 Version: 2.8
  4 Release: 1%{?dist}
  5 Source0: %{name}-%{version}.tar.bz2  <===== HERE!!!!
  6 License: GPLv3+
  7 Group: Development/Tools
```

Replace the existing bzip2 command with the pbzip2 command as an workaround on Ubuntu 16.04 X64 LTS.
```
u$> sudo apt-get install rpm pigz pbzip2 lbzip2 
u$> sudo mv /bin/bzip2 /bin/bzip2.disable
u$> sudo cp /usr/bin/pbzip2 /bin/bzip2

```
And the, let's execute "rpmbuild -bb sample.spec" command. Is it okay? 
Alternatively, I can utilize 'alias' command as following: 
```
alias tar='tar --use-compress-program=pbzip2'
alias gzip='/usr/bin/pigz'
alias gunzip='/usr/bin/unpigz'
alias bzip2='/usr/bin/pbzip2'
alias bunzip2='/usr/bin/pbunzip2'
```

**Comment by ffesti** [comment]
The main problem here is that rpm uses compression libraries for creating and unpacking the package content. Unfortunately the number of libraries offering parallel compression are very limited. There is now support for multi threaded xz compression if compiled against xz  > 5.2.0. You can set 
%_source_payload %_binary_payload to "w9T16.xzdio" with the first number being the compression level while the second being the number of threads.

External tools are all configurable by macro. Switching to more modern variants is not something that rpm upstream is likely to do soon. It is the decision of the distributions to support them and change the macros accordingly. Upstream may follow the consensus of the distributions later.

**Comment by leemgs** [comment]
> It is the decision of the distributions to support them and change the macros accordingly. 
> Upstream may follow the consensus of the distributions later.

Right. However, I hope that rpm upstream will enable multi threaded xz compression by default for them. :) For example, 

```
# https://fedoraproject.org/wiki/Features/XZRpmPayloads
%define _compression_level    7
%define _smp_mflags -j%(echo "`/usr/bin/getconf _NPROCESSORS_ONLN`")
%_source_payload   w%{_compression_level}T%{_smp_mflags}.xzdio
%_binary_payload   w%{_compression_level}T%{_smp_mflags}.xzdio
```

**Comment by pmatilai** [comment]
Multithreaded compression is not some magic holy grail. Sure it can make the compression phase faster but it requires considerably more memory and more importantly the compression ratio degrades so for distros it might be a bad choice. Also AFAICT deltarpm requires bit-by-bit equivalent compression everywhere to operate correctly and that's not the case if compression depends on thinks like number of processors on the system.

Point is, anybody considering enabling it needs to consider whether it's an actual win for a particular use-case - there *are* cases like CI builds where speed can be far more important than compression ratio.

**Comment by ffesti** [comment]
Two ways to drive the adoption of multi threaded compressors could be
a) Talk to distributions to make those compressors part of their core set of packages and change the macros in their rpm package
b) Offer patches upstream to add those compressors as alternatives in the macros.in or configure file to make it easier for distributions to switch.

**Comment by Conan-Kudo** [comment]
I would hazard to say I would _almost never_ recommend switching to multithreaded stuff by default, after accounting for how it would potentially break DeltaRPM. Unless there's a way to make _that_ deterministic, I see no pathway that would allow for widespread usage of multithreaded compression.

**Comment by pmatilai** [comment]
So...

I've yet to see any evidence that parallel *decompression* speeds up anything at all. In fact, that none of the p* variants implement that seems to suggest it doesn't. So speculating on how to best change the macros to use the parallel versions is totally moot because they're only used for decompression.

As for parallel compression, we already have that for XZ which is the most widely used format by distros these days since it offers the best compression. Switching to XZ as upstream default could be discussed (in a separate ticket/PR) but enabling parallel compression by default is not going to happen because it's a tradeoff to the worse in many common usecases.
